types/txn.go

```diff
@@ -182,13 +182,17 @@ func (ctx *TxParseContext) ParseTransaction(payload []byte, pos int, slot *TxSlo
 
 	// Remember where signing hash data begins (it will need to be wrapped in an RLP list)
 	sigHashPos := p
-	// If it is non-legacy tx, chainId follows, but we skip it
 	if !legacy {
-		dataPos, dataLen, err = rlp.String(payload, p)
+		p, err = rlp.U256(payload, p, &ctx.ChainID)
 		if err != nil {
 			return 0, fmt.Errorf("%w: chainId len: %s", ErrParseTxn, err)
 		}
-		p = dataPos + dataLen
+		if ctx.ChainID.IsZero() { // zero indicates that the chain ID was not specified in the tx.
```

Comment:
I don't think we should allow 0 chain ID for non-legacy transactions. I believe geth disallows 0 chain ID in their transaction pool even for legacy transactions and only allows pre-[EIP-155](https://eips.ethereum.org/EIPS/eip-155) transactions in incoming blocks assembled by other clients.

Reply:
@yperbasis i referred to this comment:  https://github.com/ethereum/go-ethereum/blob/c503f98f6d5e80e079c1d8a3601d188af2a899da/core/types/transaction_signing.go#L208

Reply:
> @yperbasis i referred to this comment: https://github.com/ethereum/go-ethereum/blob/c503f98f6d5e80e079c1d8a3601d188af2a899da/core/types/transaction_signing.go#L208

Yeah, but in their txpool they use `(s londonSigner) Sender`, which doesn't accept 0 chain ID, not `(s londonSigner) SignatureValues`.